docs/Threading.md

```diff
@@ -0,0 +1,167 @@
+The `_switch` routine [restores the stack pointer](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L147) of the thread being switched to and pushes the new instruction pointer value onto the new stack.
+
+If this thread is a new thread, its instruction pointer was [set during thread initialisation](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L540).
+If this thread has previously yielded, the instruction pointer is the [return address for the `_switch` call that switched away from the lthread](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L134).
+
+Note that `_switch` does *not* switch between two arbitrary lthreads.
+One of the threads involved in a `_switch` call is always the scheduler.
+The `_switch` call in `_lthread_resume` switches to another thread, the call [in `_lthread_yield_cb`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L340) and [in `_lthread_yield`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L346) switch back to the scheduler.
+
+
+Locking
+-------
+
+There are two primitives for building locks in the lthreads implementation.
+Ticket locks are spinlocks, futexes provide a fast path using atomic operations but then sleep.
+
+### Ticked locks
+
+Ticket locks are defined and documented in [src/include/enclave/ticketlock.h](src/include/enclave/ticketlock.h).
+Acquiring a ticket lock will spin in a compare and swap (CAS) loop.
+If an lthread yields while holding a ticket lock, there is a chance of deadlock because the next thread to try to acquire the lock will spin in a CAS loop and not yield.
+This may, in turn, prevent the thread holding the lock from being able to run.
+
+Ticket locks are used to implement some parts of the lthread scheduler.
+This means that there is no way for them to be made to yield in their spin loop.
+Yielding transfers control from the current lthread to the scheduler and so the concept does not exist when the lthread is running.
+
+### Lthread futexes
+
+The lthread futex implementation is a simplified version of the Linux `futex` system call.
+A futex is a 32-bit word that can be used to hold state used to implement arbitrary locking primitives.
+On the fast path, all updates to the futex word are done by CPU atomic operations.
+For example, a mutex could be implemented by treating a value of 0 as unlocked and a value of 1 as locked, with a CAS changing the state from 0 to 1 to acquire the lock on the fast path.
+The futex calls handle the slow path, where one or more threads need to wait.
+
+The `enclave_futex_wait` call suspends the thread until a corresponding `enclave_futex_wake` call on the same key.
```

Comment:
```suggestion
The `enclave_futex_wait` call suspends a given thread until a corresponding `enclave_futex_wake` call for that thread.
```

There's been no mention of "key" so while correct, I think it would lead to questions.

Reply:
```suggestion
The `enclave_futex_wait` call suspends a given thread until another thread makes a corresponding `enclave_futex_wake` call for the same futex word address, or until a timeout specified by the caller is reached.
```